File: lib/commands/add_gitignore/domain/add_gitignore_usecase.dart
import 'package:icli/di/di.dart';
import 'package:icli/utils/file_helper.dart';
import 'package:icli/utils/resources.dart';
import 'package:icli/utils/ui.dart';

class AddGitignoreUseCase {
  UI? ui;
  FileHelper? fileHelper;
  Resources? resources;

  AddGitignoreUseCase({UI? ui, FileHelper? fileHelper, Resources? resources}) {
    this.ui = get<UI>(object: ui);
    this.fileHelper = get<FileHelper>(object: fileHelper);
    this.resources = get<Resources>(object: resources);
  }

  run(bool force) {
    final ignore = resources?.gitignore();

    if (ignore != null) {
      _write('.gitignore', ignore, force);
    } else {
      ui?.error("Could not get gitignore template or class");
    }
  }

  _write(String to, String content, bool force) {
    if ((fileHelper?.fileExists(to) ?? true) && !force) {
      ui?.echo("Current dir already have gitignore file, to override it re-run the command with --f", Color.yellow);
      return;
    }

    fileHelper?.writeFile(to, content);
    ui?.echo("added gitignore", Color.cyan);
  }
}

File: icli/lib/usecase/add_gitignore/add_gitignore_usecase.rb
# frozen_string_literal: true

module ICLI
  # Add basic gitignore for new xcodeproject and swift.
  # This file will be placed if no ignore is found or overwritten if force paramenter is true.
  # It will alocate a internal copy example from inside resources.
  class AddGitignoreUsecase
    def initialize
      @ui = ServiceLocator.resolve UI
      @resources = ServiceLocator.resolve Resources
      @file_helper = ServiceLocator.resolve FileHelper
    end

    def run(force: false)
      @ignore = '.gitignore'
      force ? overwrite : create
    end

    private

    def overwrite
      @ui.echo('Overwritting gitignore', 'yellow') if gitignore?
      @file_helper.cp from: @resources.gitignore, to: @ignore
    end

    def create
      if gitignore?
        @ui.echo 'Project already have ignore. You can overwrite it with --force flag', 'yellow'
        return
      end

      overwrite
    end

    def gitignore?
      @file_helper.exists? @ignore
    end
  end
end
